### PR TITLE
NAS oneshot (stage 5) - strategy

### PR DESCRIPTION
**Pull request by ultmaster**

### CODE REVIEW

File: nni/nas/oneshot/pytorch/strategy.py [hunk 1]
```diff
@@ -12,223 +12,951 @@
...
+        :class:`~nni.nas.oneshot.pytorch.supermodule.base.BaseSuperNetModule`.
+        They are invoked in :func:`~nni.nas.oneshot.pytorch.base_lightning.traverse_and_mutate_submodules`, on each submodules.
+        For each submodule, the hook list are invoked subsequently,
+        the later hooks can see the result from previous hooks.
+        The modules that are processed by ``mutation_hooks`` will be replaced by the returned module,
+        stored in :attr:`nas_modules`, and be the focus of the NAS algorithm.
+
+        The hook list will be appended by ``default_mutation_hooks`` in each one-shot module.
+
+        To be more specific, the input arguments are four arguments:
```
Comment (QuanluZhang) [inline nni/nas/oneshot/pytorch/strategy.py:132]: whose input arguments?

File: nni/nas/oneshot/pytorch/strategy.py [hunk 2]
```diff
@@ -12,223 +12,951 @@
...
+    .. versionadded:: 2.8
+    
+       Supports searching for ValueChoices on operations, with the technique described in
+       `FBNetV2: Differentiable Neural Architecture Search for Spatial and Channel Dimensions <https://arxiv.org/abs/2004.05565>`__.
+
+    The supported mutation primitives of GumbelDARTS are:
+
+    * :class:`nni.nas.nn.pytorch.LayerChoice`.
+    * :class:`nni.nas.nn.pytorch.InputChoice`.
+    * :class:`nni.nas.nn.pytorch.ParameterizedChoice` (only when parameters are choices and type is in {supported_ops}).
```
Comment (QuanluZhang) [inline nni/nas/oneshot/pytorch/strategy.py:672]: what is the difference between ParameterizedChoice and ValueChoice?

File: nni/nas/oneshot/pytorch/strategy.py (hunk 1 above)
Comment (matluster) [reply nni/nas/oneshot/pytorch/strategy.py:132]: hooks'

File: nni/nas/oneshot/pytorch/strategy.py (hunk 2 above)
Comment (matluster) [reply nni/nas/oneshot/pytorch/strategy.py]: Typo here. It should be `ParametrizedModule`, which is a `nn.Module` with ValueChoice in its init argument.

File: nni/nas/oneshot/pytorch/strategy.py [hunk 3]
```diff
@@ -12,223 +12,951 @@
...
+
+       Supports searching for ValueChoices on operations, with the technique described in
+       `FBNetV2: Differentiable Neural Architecture Search for Spatial and Channel Dimensions <https://arxiv.org/abs/2004.05565>`__.
+       One difference is that, in DARTS, we are using Softmax instead of GumbelSoftmax.
+
+    The supported mutation primitives of DARTS are:
+
+    * :class:`nni.nas.nn.pytorch.LayerChoice`.
+    * :class:`nni.nas.nn.pytorch.InputChoice`.
+    * :class:`nni.nas.nn.pytorch.ParametrizedModule` (only when parameters are choices and type is in {supported_ops}).
```
Comment (QuanluZhang) [inline nni/nas/oneshot/pytorch/strategy.py:506]: ParameterizedModule? instead of ParametrizedModule

**Review by QuanluZhang — APPROVED** [review]

**Review by QuanluZhang — APPROVED** [review]

**Review by liuzhe-lz — APPROVED** [review]

**Pull request merged**